### PR TITLE
Populate the PK from the Singer discovery run (#2713)

### LINKED ISSUE #2713: populate primary key from singer catalog in singer_helpers.py
## Tell us about the problem you're trying to solve
Singer catalogs sometimes contain information about the primary key. We should read those and use them to populate the generated airbyte catalog in singer_helpers.py

┆Issue is synchronized with this [Asana task](https://app.asana.com/0/1200367912513076/1200368222679371) by [Unito](https://www.unito.io)

### LINKED ISSUE #5095: Test & merge contribution 4789
## Tell us about the problem you're trying to solve
The contribution in #4789 adds the ability to recognize primary key from [Singer](github.com/singer-io) catalogs. We would like to merge this, but first we need to write some unit tests for the relevant portions of the code. This may require some refactors to the code as well.

### PR DESCRIPTION
**Pull request by mildbyte**
## What
This should fix #2713 by using the Singer tap's catalog to populate `source_defined_primary_key`.

## How
Use the `key_properties` field (https://github.com/singer-io/getting-started/blob/master/docs/SPEC.md#schema-message) which is mandatory for each stream's JSONSchema and can only contain top-level fields. Something like `['user_id', 'timestamp']` will get converted to `source_defined_primary_key: [['user_id'], ['timestamp']]`.

This also allows a source to override the PK by implementing `get_primary_key_overrides()`, similar to overriding the sync mode/cursor fields.

## Note
I'm not sure what's the best way to test this, since I couldn't find a test suite for `base-singer`. I think the implication is that it will change the output catalog of _all_ Singer-backed Airbyte sources, so this PR will also need changing all those fixtures.

## Pre-merge Checklist
Expand the checklist which is relevant for this PR. 

<details><summary> <strong> Connector checklist </strong> </summary>
<p>

- [ ] Issue acceptance criteria met
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)
- [ ] Secrets are annotated with `airbyte_secret` in the connector's spec
- [ ] Credentials added to Github CI if needed and not already present. [instructions for injecting secrets into CI](https://docs.airbyte.io/contributing-to-airbyte/building-new-connector#using-credentials-in-ci). 
- [x] Unit & integration tests added as appropriate (and are passing)
    * Community members: please provide proof of this succeeding locally e.g: screenshot or copy-paste acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] `/test connector=connectors/<name>` command as documented [here](https://docs.airbyte.io/contributing-to-airbyte/building-new-connector#updating-an-existing-connector) is passing. 
    * Community members can skip this, Airbyters will run this for you. 
- [ ] Code reviews completed
- [ ] Documentation updated 
    - [ ] `README.md`
    - [ ] `docs/SUMMARY.md` if it's a new connector
    - [ ] Created or updated reference docs in `docs/integrations/<source or destination>/<name>`.
    - [ ] Changelog in the appropriate page in `docs/integrations/...`. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
    - [ ] `docs/integrations/README.md` contains a reference to the new connector
    - [ ] Build status added to [build page](https://github.com/airbytehq/airbyte/blob/master/airbyte-integrations/builds.md)
- [ ] Build is successful
- [ ] Connector version bumped like described [here](https://docs.airbyte.io/contributing-to-airbyte/building-new-connector#updating-a-connector)
- [ ] New Connector version released on Dockerhub by running the `/publish` command described [here](https://docs.airbyte.io/contributing-to-airbyte/building-new-connector#updating-a-connector)
- [ ] No major blockers
- [ ] PR merged into master branch
- [ ] Follow up tickets have been created
- [ ] Associated tickets have been closed & stakeholders notified
</p>
</details>

<details><summary> <strong> Connector Generator checklist </strong> </summary>
<p>
   
- [ ] Issue acceptance criteria met
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)
- [ ] If adding a new generator, add it to the [list of scaffold modules being tested](https://github.com/airbytehq/airbyte/blob/master/airbyte-integrations/connector-templates/generator/build.gradle#L41)
- [ ] The generator test modules (all connectors with `-scaffold` in their name) have been updated with the latest scaffold by running `./gradlew :airbyte-integrations:connector-templates:generator:testScaffoldTemplates` then checking in your changes
- [ ] Documentation which references the generator is updated as needed.
</p>
</details>

### CODE REVIEW

**Review by sherifnada — COMMENTED** [review]
@mildbyte thanks for your contribution! 


could you make this change in the `airbyte-cdk/python` package instead of here? `base-singer` is deprecated in favor of the CDK. 

The file names should be exactly the same. 

Otherwise this looks great!

**Comment by mildbyte** [comment]
@sherifnada Great, thanks -- I moved the changes to `airbyte-cdk/python` instead!

**Review by sherifnada — APPROVED** [review]
Thanks @mildbyte ! we will release this very soon. Are you blocked on using it?

**Comment by mildbyte** [comment]
@sherifnada That sounds sweet! Not really blocked -- was experimenting with the Google Search Console and GitHub sources to see if they can replace the old Singer taps we currently use and noticed that the `append_dedup` mode doesn't work with basic normalization, since there's no PK (dedup is vital for the Search Console source which always emits data a few days before the bookmark, so the PK is required there).

**Comment by sherifnada** [comment]
got it. Will release this next week.

**Comment by sherifnada** [comment]
haven't been able to get to this -- the next steps are to write test for this change and release it with the Airbyte CDK

**Review by bazarnov — APPROVED** [review]
Looks LGTM.

File: airbyte-cdk/python/CHANGELOG.md [hunk 1]
```diff
@@ -1,7 +1,10 @@
 # Changelog
 
+## 0.1.8
+Allow to fetch primary key info from singer catalog
+
```
Comment (sherifnada) [inline airbyte-cdk/python/CHANGELOG.md:5]: can you add the changelog for 1.7?

**Review by sherifnada — APPROVED** [review]
great testing

File: airbyte-cdk/python/CHANGELOG.md (hunk 1 above)
Comment (yaroslav-dudar) [reply airbyte-cdk/python/CHANGELOG.md:5]: yeap

**Pull request merged**